File: csharp/DVSClient/DVSClient/address/DatasetCombinations.cs
namespace DVSClient.Address
{
    public static class DatasetCombinations
    {
        private static readonly IDictionary<SearchType, IEnumerable<IEnumerable<Dataset>>> CombinationsDetails =
            new Dictionary<SearchType, IEnumerable<IEnumerable<Dataset>>>
            {
                {
                    SearchType.Autocomplete,
                    new List<IEnumerable<Dataset>>
                    {
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalElectricity },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalGas },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAddress, Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalGas, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalBusiness },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalElectricity, Dataset.GbAdditionalGas, Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence }
                    }
                },
                {
                    SearchType.Singleline,
                    new List<IEnumerable<Dataset>>
                    {
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNames, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalNames }
                    }
                },
                {
                    SearchType.Typedown,
                    new List<IEnumerable<Dataset>>
                    {
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalBusiness, Dataset.GbAdditionalNames, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalNames },
                        new List<Dataset> { Dataset.GbAdditionalNotyetbuilt, Dataset.GbAdditionalNames },
                    }
                },
                {
                    SearchType.Validate,
                    new List<IEnumerable<Dataset>>
                    {
                        new List<Dataset> { Dataset.GbAdditionalMultipleresidence, Dataset.GbAdditionalNotyetbuilt },
                    }
                }
            };

        public static IEnumerable<IEnumerable<Dataset>> FromSearchType(this SearchType searchType)
        {
            return CombinationsDetails[searchType];
        }
    }
}

File: typescript/src/address/datasetCombinations.ts

import { Dataset, Datasets } from "./dataset";
import { SearchType } from "./searchType";

export class DatasetCombinations {

    private static readonly CombinationsDetails: { [key in SearchType]: Dataset[][] } = {
        [SearchType.Autocomplete]: [
            [Datasets.GbAddress, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAddress, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAddress, Datasets.GbAdditionalBusiness],
            [Datasets.GbAddress, Datasets.GbAdditionalElectricity],
            [Datasets.GbAddress, Datasets.GbAdditionalGas],
            [Datasets.GbAddress, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAddress, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAddress, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAddress, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAddress, Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalBusiness],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalBusiness],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalGas, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalBusiness],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalBusiness],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalBusiness],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalBusiness],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalElectricity, Datasets.GbAdditionalGas, Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence]
        ],
        [SearchType.Singleline]: [
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNames, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalNames]
        ],
        [SearchType.Typedown]: [
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalBusiness, Datasets.GbAdditionalNames, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalNames],
            [Datasets.GbAdditionalNotyetbuilt, Datasets.GbAdditionalNames]
        ],
        [SearchType.Validate]: [
            [Datasets.GbAdditionalMultipleresidence, Datasets.GbAdditionalNotyetbuilt]
        ],
        [SearchType.LookupV2]: [],
        [SearchType.LookupV1]: []
    };

    public static fromSearchType(searchType: SearchType): Dataset[][] {
        return this.CombinationsDetails[searchType];
    }
}
